More precise check for shared storage check in inductor/reinplace pass

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147050

Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
better shared storage check in reinplace pass

[ghstack-poisoned]

laithsakka committed:
Update on "More precise check for shared storage check in inductor/reinplace pass"


Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "More precise check for shared storage check in inductor/reinplace pass"


Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
remove the opt from the list args portion to see if the failures persist or not, then will revisit .

laithsakka committed:
Update on "More precise check for shared storage check in inductor/reinplace pass"


Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "More precise check for shared storage check in inductor/reinplace pass"


Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_inductor/fx_passes/reinplace.py:
```diff
@@ -460,20 +462,36 @@ def is_meta_only_user(node):
                 return True
         return False
 
+    def _overlap(ls) -> bool:
+        try:
+            return len(compute_overlapping_tensors(ls)) != 0
+        except GuardOnDataDependentSymNode:
+            # If we fail with data dependent error we assume they all overlap.
+            return True
```
nit: "_any_tensors_overlap_conservative"?

Also this doesn't need to be a method on the class, right?

laithsakka replied:
it does not i can move it more up outside

laithsakka replied:
made it local

zou3519 commented on torch/_inductor/fx_passes/reinplace.py:
```diff
@@ -460,20 +462,36 @@ def is_meta_only_user(node):
+
     def can_inplace(node, mutated_arg):
         if isinstance(mutated_arg, (list, tuple)):
+            # TODO Using _overlap here causes a several issues.
```
How did you want to use it here?

laithsakka replied:
basically if we have list(arg0, arg1...) 
if those all do not intersect with share storage we can in-place them all.
so that way i did it is
if _overlap([arg for arg in mutated_arg]):

zou3519 reviewed (approved):
thank you

bdhirsh commented on torch/_inductor/fx_passes/reinplace.py:
```diff
@@ -460,20 +462,36 @@ def is_meta_only_user(node):
+        try:
+            return len(compute_overlapping_tensors(ls)) != 0
+        except GuardOnDataDependentSymNode:
+            # If we fail with data dependent error we assume they all overlap.
```
is `ls` guaranteed to be a list of tensors that all share the same `storage` object?

(if not, our fallback here can just be to return true if any of the tensors share the same storage, and false otherwise)

laithsakka replied:
for the current usage it will only receive things that have the same storage , mhmm yeh maybe why i got some faluires when i used it on the other location. good point i can add a comment or just harden it for future usages

laithsakka replied:
add comment

laithsakka committed:
Update on "More precise check for shared storage check in inductor/reinplace pass"


Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "More precise check for shared storage check in inductor/reinplace pass"


Currently if two tensor share storage we have some logic to avoid re-inplacing. Before this PR two tensors share storage if use same underlying storage even if they do not overlap. This diff enhance the checks to avoid cases when we know tensors do not overlap easily. 
mitigate https://github.com/pytorch/pytorch/issues/139628 but does not fix the inductor issue in it. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
rebase on main

laithsakka commented:
@pytorchbot merge
